mantidproject/mantid
# Fix problem with PaalmanPingsMonteCarloAbsorptionTest on Windows

**Description of work.**

The PaalmanPingsMonteCarloAbsorptionTest.FlatPlateTest was failing intermittently on Windows. This problem was stopping the Windows nightly build being produced on a lot of nights.

The cause of this was two things:

a) there was a problem in the duplicates check when adding points into a track in Track::addPoint which meant that sometimes a point wasn't added even though no duplicate existed already. I think this problem has perhaps come about because the track uses a boost smallvector which seems to have some slots pre-initialised so the past the end iterator end() sometimes pointed to a valid V3D point
b) there is a vector of V3D points that is populated in CSGObject::interceptSurface. The sort order of this wasn't well defined meaning that a) produced different results on subsequent runs of the same system test. The sort order is driven by the order of some pointers to Surfaces in the CSGObject::m_surList member. m_surList is created and was previously sorted in CSGObject::createSurfaceList in preparation for removing duplicate surfaces

I have fixed the bug in a) and have removed the sort on CSGObject::m_surList. Have also added a unit test on Track class that failed previously.

The fix has changed the system test reference file slightly

**To test:**

Run the PaalmanPingsMonteCarloAbsorptionTest.FlatePlateTest on Windows and verify it passes. The system tests aren't run on Windows as part of the Jenkins jobs

systemTest.bat -C Release -R PaalmanPingsMonteCarloAbsorptionTest.FlatPlateTest

Fixes #29806 .

*This does not require release notes* because **it's a bug fix which I think was introduced in some earlier 6.0 work (#29599)**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
